improve `cosc(::Float32)` and `cosc(::Float64)` accuracy

nsajko opened the pull request:
Accomplished by:

* Using minimax instead of Taylor polynomials. They're more efficient. The minimax polynomials are found using Sollya.

* Using multiple polynomials for different subintervals of the domain. Known as "domain splitting". Enables good accuracy for a wider region around the origin than possible with only a single polynomial.

The polynomial degrees are kept as-is.

Fixes #59065

nsajko committed:
a bit of code deduplication

nsajko committed:
eliminate common subexpression

nsajko committed:
better accuracy for `cosc` for `Float32`, `Float64` around `x = 0`

Improve accuracy for `cosc(::Float32)` and `cosc(::Float64)`. This is
accomplished by:

* Using minimax instead of Taylor polynomials. They're more efficient.
  The minimax polynomials are found using Sollya.

* Using multiple polynomials for different subintervals of the domain.
  Known as "domain splitting". Enables good accuracy for a wider region
  around the origin than possible with only a single polynomial.

The polynomial degrees are kept as-is.

Fixes #59065

nsajko commented:
To clarify: the above plots show the error in ULPs. In each case, the plot happens after downsampling by taking the maximum. Also see the linked issue.

nsajko committed:
test accuracy

nsajko committed:
add copyright notice

oscardssmith commented on test/testhelpers/ULPError.jl:
```diff
@@ -0,0 +1,57 @@
+        throw(ArgumentError("invalid"))
+    end
+    function ulp_error(accurate::AbstractFloat, approximate::AbstractFloat)
+        if isnan(accurate)
+            @noinline throw_invalid()
+        end
+        if isnan(approximate)
+            @noinline throw_invalid()
+        end
```
this should return `Inf` if one is NaN and the other isn't, or 0 if both are NaN.

nsajko committed:
ULPError simplification as per Oscar's suggestion

nsajko commented:
Test failure seems surely unrelated.

nsajko committed:
ULPError common subexpression elimination

nsajko committed:
ULPError: handle `isinf(approximate)`

nsajko committed:
put comment in correct position

nsajko committed:
style: get rid of the `let`s/nesting

nsajko committed:
style: simplify `return`s

nsajko committed:
test `ULPError` edge cases

oscardssmith commented:
I've read the pr 3 time by now, and I'm still unsure whether the struct it introduces is worth it. I feel like I n better read the more function based approach in exp.jl, but some portion of that is that I'm the author of those.

nsajko committed:
fix `ULPError` edge case regarding zero/inf

nsajko committed:
`ULPError`: simplify

nsajko committed:
`ULPError`: delete dead code

oscardssmith commented on test/testhelpers/ULPError.jl:
```diff
@@ -0,0 +1,59 @@
+            else
+                inf_return
+            end
+        end
```
maybe put all of this inside `!(isfinite(approx) && isfinite(accurate))`? Also, I don't think these variables are used enough to be worthwhile.

oscardssmith commented on test/testhelpers/ULPError.jl:
```diff
@@ -0,0 +1,59 @@
+            end
+        end
+        # assuming `precision(BigFloat)` is great enough
+        acc = if accurate isa BigFloat
+            accurate
+        else
+            BigFloat(accurate)::BigFloat
+        end
```
why is this necessary? I would trust the precision of `accurate` It's often nice to be able to use lower precision than BigFloat to avoid the allocation and speed penalty.

nsajko replied:
My idea is that `ULPError` will only be used for tests, so accuracy trumps performance.

oscardssmith replied:
`Float64` is fine, and `QuadMath` is super duper good enough for the rest of the cases. Even for testing, it is quite nice to be able to quickly do a couple billion tests.

nsajko committed:
`ULPError`: eliminate variables only used once

nsajko committed:
`ULPError`: use `Float32` literals

nsajko committed:
`ULPError`: guard the edge case handling behind a single `if`

nsajko committed:
`ULPError`: use `Float64` instead of `BigFloat`

nsajko committed:
`ULPError`: avoid unnecessary conversion to `BigFloat`

nsajko committed:
delete debugging throw

Should not be necessary now we have some tests in the test suite.

nsajko committed:
test: add missing edge case

nsajko committed:
test: `ULPError`: faitfhul approximation

nsajko committed:
Revert "`ULPError`: avoid unnecessary conversion to `BigFloat`"

This reverts commit 33eca5fefbf48e94b6d26a533d03dc8d481ac247.

nsajko committed:
handle precision better

Trust the precision of `accurate` is great enough, unless it's one of
the known-imprecise types.

nsajko committed:
expand `ULPError` test suite

nsajko committed:
`ulp_error_maximum`: use `Fix1`

As suggested by devmotion on a PR to LogExpFunctions.jl:

* https://github.com/JuliaStats/LogExpFunctions.jl/pull/99

nsajko committed:
use `ulp_error_maximum` instead of a loop

As suggested by devmotion on a PR to LogExpFunctions.jl:

* https://github.com/JuliaStats/LogExpFunctions.jl/pull/99

nsajko committed:
`ULPError`: delete unnecessary method static parameters

Specialization on the callable argument types should happen anyway.

nsajko committed:
improve style

stevengj commented:
Also, you might want to check that `cosc(0.0) === -0.0` and `cosc(-0.0) === +0.0`.

nsajko commented:
> is this the minimax polynomial that minimizes the maximum absolute error over the interval, or which minimizes the maximum relative error?

It minimizes relative error. The [`fpminimax`](https://www.sollya.org/sollya-8.0/help.php?name=fpminimax) commands that appear in the included Sollya script allow choosing either absolute or relative error to optimize for, but defaults to the `relative` option.

Anyway if I had chosen to minimize the absolute error with the polynomial, the ULP error plots would have looked quite different.

> I'm a bit concerned that your plot seems to show 1ulp error at x == 0. What does that mean? A nonzero error at the root would correspond to an infinite relative error.

At `x == 0`, the result is exact (easy to check exhaustively, see a relevant separate PR inspired by your comments: #59265). Otherwise the relative error would not be finite.

Each point on the plot is the result of many evaluations in the same neighborhood, which are aggregated basically by taking the maximum value. Trying to plot every evaluation would result in an unreadably noisy plot, so the idea is basically to get rid of the downward spikes of the noise. Thus the plot visualizes the maximum error in the neighborhood of each point, but not the minimum error. The justification is that the worst case is exactly what is interesting.

As an additional explanation, I'll copy over a paragraph from my message on this PR:

* https://github.com/JuliaStats/LogExpFunctions.jl/pull/101:

> The plotting happens after downsampling (which just takes the maximum value), which itself happens after smoothing (which also just takes the maximum, i.e., a sliding window maximum). Thus the downsampling and smoothing basically just remove the downward spikes of the noise, preserving the maximum error at the region around each point.

If you want I can also share the code I use for plotting. The ULP error itself is calculated using the same code included in this PR, see module `ULPError`. The downsampling is implemented using some iterators, at least some of which I want to eventually include into IterTools.jl. The initial PR (implementing the sliding window maximum, used for smoothing, which happens before downsampling proper) is:

* https://github.com/JuliaCollections/IterTools.jl/pull/126

> Since the `x == 0` root is known exactly, and is exactly representable, we should strive to have small relative error as we approach it.

IMO it is fine to minimize the maximum relative error over the entire interval, there's no need to treat the immediate neighborhood of the root specially. (In fact, Sollya is not able to do that in the first place. My own package FindMinimaxPolynomial *is* able to do that, however it lacks Sollya's functionality for optimizing the coefficients of the polynomial in machine precision.)

> Ensuring this is a nice property of using a Taylor series around a zero.

An exhaustive comparison of the maximum ULP errors between zero and `nextfloat(type(0), 100_000_000)` before and after this PR shows the maximum error happens to stays the same in this immediate neighborhood of zero. They are `0.58824617f0` for `Float32` and `0.49999997f0` for `Float64`.

> I'm also not entirely clear on what precisely you are plotting. I'm imagining a plot of relative error |approx–exact|/|exact|, divided by eps(T) to get something ulp-like?

See the module `ULPError` in `test/testhelpers/ULPError.jl`. In particular, the generic formula is:

```julia
abs((approximate - acc) / eps(approximate))
```

I believe this corresponds exactly to one of the definitions of ULP error, so it's not merely "ULP-like"?

nsajko committed:
get rid of the `struct`

nsajko committed:
fix

nsajko committed:
merge two methods into one

StefanKarpinski commented:
@stevengj, if that addresses your concerns and @oscardssmith is happy, can we merge this?

oscardssmith commented:
I don't think `sinc`/`cosc` should exist in `Base`, but while they do, they might as well be good. Approved.
